sys/xtimer/xtimer.c

```diff
@@ -154,20 +155,20 @@ static int _msg_wait(msg_t *m, msg_t *tmsg, xtimer_t *t)
 }
 
 int _xtimer_msg_receive_timeout64(msg_t *m, uint64_t timeout_ticks) {
-    msg_t tmsg;
-    xtimer_t t;
-    _setup_timer_msg(&tmsg, &t);
-    _xtimer_set_msg64(&t, timeout_ticks, &tmsg, sched_active_pid);
+    msg_t tmsg = { .sender_pid  = sched_active_pid,
+                   .type        = MSG_XTIMER,
+                   .content.ptr = &tmsg };
+
+    xtimer_t t = { .callback = _callback_timeout_msg,
+                   .arg      = &tmsg };
+
+    _xtimer_set64(&t, timeout_ticks, timeout_ticks >> 32);
```

Comment:
This is now racey.

If the thread gets scheduled away after this set(), the timer triggers, then the thread gets scheduled again, it'll sit in msg_receive() forever.

Reply:
Good catch! Ok how about something like the idea I sketched in [#13429 comment ](https://github.com/RIOT-OS/RIOT/pull/13429#issuecomment-589673697) then?